cis194/week4/zehua/src/Exercises.hs

```diff
@@ -0,0 +1,27 @@
+module Exercises where
+
+import Data.Bool (bool)
+
+-- ex1
+fun1 :: [Integer] -> Integer
+fun1 [] = 1
+fun1 (x:xs)
+  | even x = (x - 2) * fun1 xs
+  | otherwise = fun1 xs
+
+fun1' :: [Integer] -> Integer
+fun1' = product . fmap (subtract 2) . filter even
+
+fun2 :: Integer -> Integer
+fun2 1 = 0
+fun2 n | even n = n + fun2 (n `div` 2)
+       | otherwise = fun2 (3 * n + 1)
+
+fun2' :: Integer -> Integer
+fun2' = foldr (+) 0
+      . filter even
+      . takeWhile (>1) -- >1 instead of /=1 to defend against negative numbers
+      . iterate (bool <$> oddv <*> evenv <*> even)
```

Comment:
We can also write the below, you are the judge to say which one is better
```
      . iterate (bool . oddv <*> evenv <*> even)
```

Reply:
@shineli1984 from what you said a generic `fmap` / `<$>` is better than the binding?

```haskell
instance Functor ((->) r) where
    fmap = (.)
```

So I guess `<$>` is better in this case?

Reply:
`... <$> ... <*> ... <*>` is a very common pattern when using applicative. 
But the choice of using `.` or `<$>` for function composition is open, at least from my point of view. Because one can argue that when apply a value on `(bool <$> oddv <*> evenv <*> even)` that value will go to `oddv` first instead of `bool` as if we are adding redundant brackets like this: `(bool <$> oddv) <*> evenv <*> even`. This is to express composition first instead of `fmap`. They have different intent but same result. So either is fine.